tests/subsys/power/power_mgmt_soc/src/power_mgmt.c

```diff
@@ -0,0 +1,316 @@
+/*
+ * Copyright (c) 2020 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <zephyr.h>
+#include <device.h>
+#include <soc.h>
+#include <power/power.h>
+#include <logging/log.h>
+#include <sys/util.h>
+#define LOG_LEVEL LOG_LEVEL_DBG
+LOG_MODULE_REGISTER(pwrmgmt_test);
+
+#define SLP_STATES_SUPPORTED      2ul
+
+/* Thread properties */
+#define TASK_STACK_SIZE           1024ul
+#define PRIORITY                  K_PRIO_COOP(5)
+
+/* Sleep time should be lower than CONFIG_SYS_PM_MIN_RESIDENCY_SLEEP_1 */
+#define THREAD_A_SLEEP_TIME       100ul
+#define THREAD_B_SLEEP_TIME       1000ul
+
+/* Maximum latency should be less than 300 ms */
+#define MAX_EXPECTED_MS_LATENCY   500ul
+
+/* Sleep some extra time than minimum residency */
+#define DP_EXTRA_SLP_TIME         1100ul
+#define LT_EXTRA_SLP_TIME         500ul
+
+#define SEC_TO_MSEC               1000ul
+
+K_THREAD_STACK_DEFINE(stack_a, TASK_STACK_SIZE);
+K_THREAD_STACK_DEFINE(stack_b, TASK_STACK_SIZE);
+
+static struct k_thread thread_a_id;
+static struct k_thread thread_b_id;
+
+struct pm_counter {
+	uint8_t entry_cnt;
+	uint8_t exit_cnt;
+};
+
+/* Track time elapsed */
+static int64_t trigger_time;
+static bool checks_enabled;
+/* Track entry/exit to sleep */
+struct pm_counter pm_counters[SLP_STATES_SUPPORTED];
+
+static void pm_latency_check(void)
+{
+	int64_t latency;
+	int secs;
+	int msecs;
+
+	latency = k_uptime_delta(&trigger_time);
+	secs = (int)(latency / SEC_TO_MSEC);
+	msecs = (int)(latency % SEC_TO_MSEC);
+	LOG_INF("PM sleep entry latency %d.%03d seconds", secs, msecs);
+
+	if (secs > 0) {
+		LOG_WRN("Sleep entry latency is too high");
+		return;
+	}
+
+	if (msecs > MAX_EXPECTED_MS_LATENCY)
+		LOG_WRN("Sleep entry latency is higher than expected");
+}
+
+/* Hooks to count entry/exit */
+void sys_pm_notify_power_state_entry(enum power_states state)
+{
+	if (!checks_enabled) {
```

Comment:
why do you need  this variable / check ?

Reply:
This is to ensure the counters are incremented only when the system goes to sleep/deep sleep caused by the two test functions. Only in the test_pwr_mgmt_singlethread and test_pwr_mgmt_multithread will the flag be assigned true so then the counters can be incremented.